docs/proposals/hugepages.md

```diff
@@ -0,0 +1,154 @@
+
+The proposal is also limited to support on x86_64 where two huge page sizes are supported: 2MB and 1G.  The design, however, should accommodate addition huge page sizes available on other architectures.
+
+**NOTE: This design, as currently proposed, requires the use of pod-levels cgroups, which are currently not available but are under development by @derekwaynecarr**
+
+## Background
+
+Huge pages are a hardware feature designed to reduce pressure on the Translation Lookaside Buffer (TLB)   The TLB is a small hardware cache of virtual-to-physical page mappings.  If the virtual address passed in a hardware instruction can be found in the TLB, the mapping can be determined quickly.  If not, a TLB miss occurs and the hardware must walk the in-memory page table to discover a physical mapping for the virtual address.
+ 
+Take a program that operates on a large 2MB structure as an example.  If the program accesses that space in such a way that one byte in each regular 4k page is accessed, 2M/4k = 512 TLB entries are needed to map the address range.  Each TLB miss results in an expensive walk of the page table.  However, if the allocation is backed by a 2M huge page by, only 1 TLB entry is required resulting in a highly likelihood that entry will remain in the cache and hit on accesses to the entire 2M structure.
+ 
+On x86_64, there are two huge page sizes: 2MB and 1G.  1G huge pages are also called gigantic pages.  1G must be enabled on kernel boot line with ```hugepagesz=1G```. Huge pages, especially 1G ones, should to be allocated early before memory fragments (i.e. at/near boot time) to increase the likelihood that they can be allocated successfully with minimal memory migration (i.e. defreg) required.
+
+## Use Cases
+
+The class of applications that benefit from huge pages are typically have
+- A large memory working set
+- A sensitivity to memory access latency
+
+Example applications include:
+- Java applications can back the heap with huge pages using the ```-XX:+UseLargePages``` option.  
+- In-memory databases
+
+Applications can generally use huge pages by calling
+- ```mmap()``` with ```MAP_ANONYMOUS | MAP_HUGETLB``` and use it as anonymous memory
+- ```shmget()``` with ```SHM_HUGETLB``` and use it as a shared memory segment (see Known Issues).
+
+### Pod Specification
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+...
+    resources:
+      requests:
+	    hugepages: "10"
+      limits:
+	    hugepages: "10"
+  nodeSelector:
+    kubernetes.io/huge-page-size: "2M"
+```
+
+Huge pages can not be overcommitted on a node.
+
+While a system may support multiple huge pages sizes, it is assumed that nodes configured with huge pages will only use one huge page size, namely the default page size in ```cat /proc/meminfo | grep Hugepagesize```.  In Linux, this is 2M unless overridden by ```default_hugepagesz=1g``` in the kernel boot parameters.
```

Comment:
```hugepagesz=1g``` enables 1G pages, but doesn't make them the default IIUC.  I'll go try it out.

https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt